[LLDB] Add type casting to DIL, part 2 or 3

cmtice opened the pull request:
This PR implements the actual type casting part. With this, type casting to builtin types should work. The third PR, which will be put up after this one is merged, will expand the type name parsing to allow casting to user-defined types.

cmtice committed:
[LLDB] Add type casting to DIL, part 2 or 3

This PR implements the actual type casting part. With this, type
casting to builtin types should work. The third PR, which will be
put up after this one is merged, will expand the type name parsing
to allow casting to user-defined types.

cmtice committed:
Fix clang format issues.

adrian-prantl commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -21,6 +21,42 @@
+                           bool use_synthetic) {
+  Status error;
+  if (!in_valobj_sp) {
+    error = Status("invalid value object");
```
is this even used?
Should the function return `llvm::Expected<ValueObjectSP>` instead?

cmtice committed:
Addressed reviewer comments:
 - Changed enum name to CastPromotionKind
 - Added Doxygen comment for VerifyTypeCast function
 - Removed unneeded Status variable
 - Rewrote VerifyTypeCast to use early style returns.

cmtice commented on lldb/include/lldb/ValueObject/DILEval.h:
```diff
@@ -71,6 +71,11 @@ class Interpreter : Visitor {
                   std::shared_ptr<ExecutionContextScope> ctx,
                   const IntegerLiteralNode *literal);
 
+  llvm::Expected<CompilerType>
```
Comment added. Yes all the reference parameters are 'out' parameters, i.e. the function can update their values.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +776,236 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+
+  } else if (target_type.IsReferenceType()) {
+    // Cast to a reference type.
+    cast_kind = CastKind::eReference;
```
I've tried to do so.

Michael137 commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -46,6 +46,13 @@ enum class CastKind {
+enum CastPromotionKind {
+  eArithmetic,
+  ePointer,
+  eNone,
```
Can we comment on the individual enum cases what they represent. Maybe giving a small `C`/`C++` expression as an example

Michael137 commented on lldb/test/API/commands/frame/var-dil/expr/Casts/main.cpp:
```diff
@@ -0,0 +1,52 @@
+#include <cstdlib>
+#include <limits>
+#include <memory>
+#include <string>
```
Can we remove some of the unused headers (e.g., `<string>`)

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -21,6 +21,37 @@
+                           lldb::DynamicValueType use_dynamic,
+                           bool use_synthetic) {
+  if (!in_valobj_sp)
+    return in_valobj_sp;
```
```suggestion
    return nullptr;
```

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -21,6 +21,37 @@
+    return in_valobj_sp;
+
+  lldb::ValueObjectSP value_sp = in_valobj_sp;
+  Target *target = value_sp->GetTargetSP().get();
```
We don't seem to be using `target` for anything

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +771,242 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+  switch (cast_kind) {
+  case CastKind::eEnumeration: {
+    if (!target_type.IsEnumerationType()) {
+      std::string errMsg = "invalid ast: target type should be an enumeration.";
```
Can we remove the prefix from the error? The callers of this function can choose to prepend such prefix if needed
```suggestion
      std::string errMsg = "target type should be an enumeration";
```

Also applies to the errors below

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +771,242 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+        "invalid ast: operand is not convertible to enumeration type";
+    return llvm::make_error<DILDiagnosticError>(m_expr, std::move(errMsg),
+                                                node->GetLocation());
+    // return error.ToError();
```
```suggestion
```

Michael137 reviewed (commented):
Gave it an initial pass

cmtice committed:
Addressed review comments & simplified code:
 - Combined cast enumeration kinds into a single enumeration
 - Reorganized code so VerifyCastKind does not update any of its parameters
 - VerifyCastKind now returns type of cast to perform (cast kind enum)
 - Removed redundant error checking from the Visit function for cast nodes.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +771,242 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+        operand_sp->Cast(target_type.GetNonReferenceType()));
+  }
+  case CastKind::eNone: {
+    switch (promo_kind) {
```
Now that I've combined the enums there is no inner switch statement.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +771,242 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+    lldb::ValueObjectSP operand_sp(
+        GetDynamicOrSyntheticValue(operand, m_use_dynamic, m_use_synthetic));
+    return lldb::ValueObjectSP(
+        operand_sp->Cast(target_type.GetNonReferenceType()));
```
If I don't get the non-reference type, then I end up with strange (and/or incorrect) results. Things like:

(lldb) v &(int&)arr_1d                                                         
(int &*) &arr_1d = 0x00007fffffffddc0
notice the odd "int &*" type

or 
(lldb) v (int&)arr_1d                                                          
(int &) arr_1d = 0x0000000200000001 (&arr_1d = <read memory from 0x200000001 failed (0 of 4 bytes read)>) 
 
versus when I get the non-reference type:
(lldb) v (int&)arr_1d 
(int) arr_1d = 1

Michael137 replied:
Ok thanks for investigating.

Can we remove the "cast to reference type" support for now then. Until we fix `ValueObject::Cast` (or confirm that it does the right thing, in which case we have to re-assess what to do in DIL casts).

That way we can have a more targeted discussion. Casting to reference types seems like a more niche use-case than pointer/scalar casts, so omitting it for now wouldn't be a big deal (in my opinion).

Michael137 commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -40,12 +40,21 @@ enum class UnaryOpKind {
+//  ePointer,    ///< Casting to a pointer type.
+//  eNone,       ///< Invalid promotion type (results in error).
+//};
+
```
Is this meant to be here?

Michael137 commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +764,182 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+Interpreter::VerifyCastType(lldb::ValueObjectSP operand, CompilerType op_type,
+                            CompilerType target_type, int location) {
+
+  if (target_type.IsScalarType()) {
```
Can we put everything that's inside this `target_type.IsScalarType()` if-block into a helper. Call it, e.g., `VerifyArithmeticCastType` or something

Michael137 reviewed (commented):
Looks better, left some more comments

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +764,182 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
 
+llvm::Expected<CastKind>
+Interpreter::VerifyCastType(lldb::ValueObjectSP operand, CompilerType op_type,
+                            CompilerType target_type, int location) {
```
I pass is separately, because (in the caller) we've checked to see if the operand's type is a reference type, and if it is we've put the non-reference type into op_type, to use in the various type-cast checks. But we don't want to change the actual type of the operand.

I will rename it to be 'source_type' as requested.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +764,182 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+  if (target_type.IsNullPtrType()) {
+    // Cast to nullptr type.
+    bool is_signed;
+    if (!target_type.IsNullPtrType() &&
```
Good catch! That was supposed to be looking at the source type, making sure it's compatible.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -740,16 +764,182 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+            op_type.TypeDescription().length());
+      }
+      // Casting pointer to bool is valid. Otherwise check if the result type
+      // is at least as big as the pointer size.
```
Will do.

cmtice commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -99,8 +122,9 @@ Interpreter::UnaryConversion(lldb::ValueObjectSP valobj, uint32_t location) {
+  llvm::StringRef name = "result";
   if (in_type.IsArrayType())
-    valobj = ArrayToPointerConversion(*valobj, *m_exe_ctx_scope);
+    valobj = ArrayToPointerConversion(*valobj, *m_exe_ctx_scope, name);
```
Accepted.

cmtice committed:
Address review comments:
  - Put arithmetic cast verification into a separate helper function.
  - Rename 'op_type' to 'source_type'.
  - Return errors if we can't get the byte sizes for the types (when we try)
  - Use early-return for ptr-to-boolean cast.
  - Fix typo in NullPtr cast checking conditions.
  - Remove unused header includes from test.

cmtice commented:
I believe I have addressed all the review comments and this is ready for another look. (Thank you!)

cmtice committed:
Fix format errors.

cmtice committed:
Fix clang-format issue.

Michael137 reviewed (commented):
Left two follow-up questions:
* https://github.com/llvm/llvm-project/pull/170332/changes#r2602221008
* https://github.com/llvm/llvm-project/pull/170332/changes#r2627412450

And some stylistic comments.

Once we resolve these we're good to go

cmtice committed:
Remove code for casting TO a nullptr type.

cmtice committed:
Fix clang-format issues.

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-typecast-part2

Michael137 reviewed (approved):
Thanks for addressing all the comments.

Can we remove the cast-to-reference-type support for now? (see my [latest comment](https://github.com/llvm/llvm-project/pull/170332#discussion_r2658206585))

We can file an GitHub issue to track/discuss further. But at least we'll have the common kinds of casts landed (scalar/pointer casts).

Once that's done, LGTM!

cmtice committed:
Fix some merge conflicts that were previously missed.

cmtice committed:
Remove casting to reference types.
